### PR TITLE
aug 2023 api updates

### PR DESCRIPTION
**Pull request by fivetran-reneeli**
## PR Overview
**This PR will address the following Issue/Feature:**
https://github.com/fivetran/dbt_iterable_source/issues/27
**This PR will result in the following new package version:** v0.11.0

Removal of fields/ models and addition of fields, especially a new primary key that will avoid fanouts caused by the previous key. 

**Please detail what change(s) this PR introduces and any additional information that should be known during the review of this PR:**

- Because email is no longer a primary key, previous joins using this field were resulting in fanouts. We have replaced it with a new PK. In addition we've updated any grains that were grouped by email and any partitions on email. Updated the unique keys in the incremental config, and the tests in the yml. 
- Updated the seed data to contain new fields brought downstream
- Added macro `does_table_exist` to check for whether unsub_* tables are history vs non-history. The end model `iterable__user_unsubscriptions` got a hefty update in response to this to accommodate the additional conditional statements
- I've updated the readme with a notice to make sure both are either history or non, in order for the logic to work ^
- Removed `user_device` related fields since we removed that staging model. 

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt compile
- [x] dbt run –full-refresh
- [x] dbt run
- [x] dbt test
- [x] dbt run –vars (if applicable)

 iterable__using_user_unsubscribed_message_type: False
 iterable__using_campaign_suppression_list_history: true

Before marking this PR as "ready for review" the following have been applied:
- [x] The appropriate issue has been linked and tagged
- [x] You are assigned to the corresponding issue and this PR
- [x] BuildKite integration tests are passing

### Detailed Validation
Please acknowledge that the following validation checks have been performed prior to marking this PR as "ready for review":
- [x] You have validated these changes and assure this PR will address the respective Issue/Feature.
- [x] You are reasonably confident these changes will not impact any other components of this package or any dependent packages.
- [x] You have provided details below around the validation steps performed to gain confidence in these changes.

Ensure the new unique/not null tests pass and the new fields persist

### Standard Updates
Please acknowledge that your PR contains the following standard updates:
- Package versioning has been appropriately indexed in the following locations:
    - [x] indexed within dbt_project.yml
    - [x] indexed within integration_tests/dbt_project.yml
- [x] CHANGELOG has individual entries for each respective change in this PR
    
- [x] README updates have been applied (if applicable)
    
- [ ] DECISIONLOG updates have been updated (if applicable)
- [x] Appropriate yml documentation has been added (if applicable)

### dbt Docs
Please acknowledge that after the above were all completed the below were applied to your branch:
- [ ] docs were regenerated (unless this PR does not include any code or yml updates)

### If you had to summarize this PR in an emoji, which would it be?
 
:dancer:

### CODE REVIEW

File: CHANGELOG.md
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:1]: ??? No CHANGELOG?

File: README.md
Comment (fivetran-joemarkiewicz) [inline README.md:93]: Same comment from the source, this shouldn't be it's own step and should rather be a component of the pre-req section.

File: macros/does_table_exist.sql
Comment (fivetran-joemarkiewicz) [inline macros/does_table_exist.sql:17]: This is not needed if it already exists in the source. Let's remove this macro here and just rely on the source.

File: models/iterable__events.sql
Comment (fivetran-joemarkiewicz) [inline models/iterable__events.sql:3]: Can you elaborate on why there are two unique keys here?

File: models/iterable__user_campaign.sql
Comment (fivetran-joemarkiewicz) [inline models/iterable__user_campaign.sql:38]: Can you elaborate on the need for this where clause?

File: packages.yml
Comment (fivetran-joemarkiewicz) [inline packages.yml:7]: Reminder to update

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-reneeli please see my comments below for change requests and questions.

Additionally, I noticed with the non history (the new) schema there are tests that do not pass.
image

Further, there are quite significant changes that are being applied to all the models and I don't see any additional validations to prove the outputs are not being affected or are what we would expect. If you have any Hex validations or other proof that these changes are sufficient that would be preferred before I progress in the PR review. I am a bit cautious that the changes do not impact the end results as I can see the above test failing.

Please address the source review notes, below comments, and also provide validations for the end models modified in this PR to confirm we are seeing the expected results before marking this ready for second review.

File: models/iterable__events.sql
Comment (fivetran-reneeli) [reply models/iterable__events.sql:3]: For an event-grain model in iterable, the unique key is `_fivetran_id` + `_fivetran_user_id` (in our model, `_fivetran_id` = `event_id`  and ` _fivetran_user_id` = coalesce( `unique_user_key` ,`email` ). 

In the old schema, only `_fivetran_id` (= `event_id` ) is the unique key.

But I reasoned this was still backwards compatible

File: models/iterable__user_campaign.sql
Comment (fivetran-reneeli) [reply models/iterable__user_campaign.sql:38]: My reasoning is, since this is a user-grain model, I wanted to exclude where there's no user associated. Since this comes from event-based model `iterable__events` rather than a user-based model, perhaps we come across events without a user. though theoretically, each event has a user associated, so `unique_user_key` should never be null. But I felt this was safe to do just in case.

File: README.md
Comment (fivetran-reneeli) [reply README.md]: updated

File: macros/does_table_exist.sql
Comment (fivetran-reneeli) [reply macros/does_table_exist.sql]: true, deleted

File: CHANGELOG.md
Comment (fivetran-reneeli) [reply CHANGELOG.md:1]: updated

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-reneeli thanks for applying the previous requested changes. I do still have some follow up questions below. Would you be able to address the remaining questions below. Let me know if you have any questions.

Additionally, the Hex validations notebook was a great start to show validations between the old and new design of the package. However, would you be able to provide additional validations for each of the impacted end models. It seems only a few were provided. 

I want to make sure that the row counts in particular are what we would expect following these updates. Please refer to the Intercom validations in the Hex validations catalog as an example of what I would be looking for in terms of row count validations. Additionally, I would like to confirm that the changes to any incremental models are running as expected in subsequent runs.

Lastly, I tested this package using both the old API and the new API versions. It seems the old API runs as expected; however, there is a test failure when using the new API `iterable_schema: iterable_user_id_20`. This seems to be the same test failure I shared in the previous review. Would you be able to address this test failure and update the package accordingly.

image

File: models/iterable__user_campaign.sql
Comment (fivetran-reneeli) [reply models/iterable__user_campaign.sql:38]: Digging more into the docs from these event tracking platforms, I'm pretty sure all events have an associated user. I believe even if the user is unknown when an event is collected, there will be an anonymous id placed. I suppose we can remove this, it was more of a just-in-case since I found it interesting that it's a user-based model but pulls from an event source. I'll remove this

File: models/iterable__events.sql
Comment (fivetran-reneeli) [reply models/iterable__events.sql:3]: From the dbt docs, I'm not sure if there's a difference between having 2 fields here versus creating a composite key in the model to use. Is the concern that this way will affect compute more drastically?

I updated this to follow the underlying unique key changes, but haven't tested it substantially. I will need to update/add/delete records in the test data to thoroughly test it.

Previously the delete+insert strategy would affect a row with matching event_id, but now that row won't be affected unless the unique_user_key field matches that of the incoming record as well.

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-reneeli thanks for working on the validations, I am feeling a lot better about the incremental model updates and ensuring they are working as expected.

I do still have some additional notes and it seems there may have been some previous comments that were missed from my last review that I would like you to address before this will be good to approve.

File: models/iterable__user_unsubscriptions.sql
Comment (fivetran-reneeli) [reply models/iterable__user_unsubscriptions.sql:60]: Oh shoot didn't see this. I had removed this due to the latest_batch_index filter yes, but now that I think about it, I suppose if a customer's using the history tables, it still would be helpful to know the updated_at time. So I'll move this back in. Thanks for noticing this!

File: README.md
Comment (fivetran-reneeli) [reply README.md:115]: sorry to miss this. Updated this portion and highlighted where to expect the pass through fields

File: models/intermediate/int_iterable__list_user_history.sql
Comment (fivetran-reneeli) [reply models/intermediate/int_iterable__list_user_history.sql:12]: So this was intentionally left since this pertains to email list ids, I figured there must be associated email fields.

File: models/intermediate/int_iterable__list_user_history.sql
Comment (fivetran-reneeli) [reply models/intermediate/int_iterable__list_user_history.sql:6]: Ah looking at the source list_user_history table, it does look like _fivetran_id replaces email. So when reproducing this table, you're right that email may not exist. I think I leant too heavily on the fact we're lagging on email_list_id, lept to conclusions here. I will remove the email filter.

File: models/intermediate/int_iterable__list_user_history.sql
Comment (fivetran-reneeli) [reply models/intermediate/int_iterable__list_user_history.sql:12]: Actually, I will change this to partition by unique_user_key. I am not sure there's a case where theres email_list_id without an email (I don't see it in our test data), but I see how it's safer to use unique_user_key

File: models/intermediate/int_iterable__list_user_history.sql
Comment (fivetran-reneeli) [reply models/intermediate/int_iterable__list_user_history.sql:33]: Ohhh. Thank you for pointing this out. I do feel there's value for customers here, especially if they're adding additional user details, I think they would expect to see them in the end iterable__users model. 

But you make a good callout that it is clunky for them to maintain. I'll veer on the side of keeping this passthrough functionality in, but make the notes in the changelog and readme

File: models/iterable__events.sql
Comment (fivetran-reneeli) [reply models/iterable__events.sql]: Following our convo, I switched to using a composite key to keep in line with standards

**Review by fivetran-joemarkiewicz — APPROVED** [review]
Looks good to me! @fivetran-reneeli thanks for working through all of these changes!! I have just two last comments to add to the CHANGELOG and then reminder to update packages.yml. This is good to go!

**Pull request merged**
